### ISSUE TITLE
Nextcloud creation - database creation fails - ./prepare_site.yml - mysql only

### ISSUE THREAD

**Issue by james-cook** [ReinerNippes/selfhosted_on_docker#3]
I'm getting a failure when I call ./prepare_site.yml for nextcloud:

> TASK [docker_nextcloud : testnc5 | first setup nextcloud] ****************************************************************************************
> Wednesday 05 February 2020  21:19:10 +0000 (0:00:00.076)       0:01:53.935 ****
> fatal: [localhost]: FAILED! => {"changed": true, "cmd": "docker exec --user www-data testnc5 php occ  maintenance:install  --database mysql --database-host testnc5-db --database-name nextcloud  --database-table-prefix oc_ --database-user nextcloud  --database-pass 1P2aMOLiulaii62puB3xX2YKRWTXXmfH  --admin-user admin --admin-pass cVmLoyLJ09V9LJD7DIVtNr7UD81oYNl5 --data-dir /var/nc-data\n", "delta": "0:00:01.163908", "end": "2020-02-05 21:19:11.943165", "msg": "non-zero return code", "rc": 1, "start": "2020-02-05 21:19:10.779257", "stderr": "", "stderr_lines": [], "stdout": "Error while trying to create admin user: Failed to connect to the database: An exception occurred in driver: SQLSTATE[HY000] [1044] Access denied for user 'nextcloud'@'%' to database 'nextcloud'\n -> ", "stdout_lines": ["Error while trying to create admin user: Failed to connect to the database: An exception occurred in driver: SQLSTATE[HY000] [1044] Access denied for user 'nextcloud'@'%' to database 'nextcloud'", " -> "]}
> 
> 

To debug this I entered the testnc5 container: 

`docker exec -it testnc5 sh`

and added the mysql client:

`apk add mysql-client`

The "equivalent" mysql to the **occ** command above (I suppose...) would be:

```
/var/www/html # mysql -u nextcloud -p'1P2aMOLiulaii62puB3xX2YKRWTXXmfH' \
>     -h testnc5-db  \
>     -D nextcloud

which gives:

````ERROR 1044 (42000): Access denied for user 'nextcloud'@'%' to database 'nextcloud'
`
```
This looks like the same error message as above (**occ** command)
I can log in using the password in _**testnc5_mysql_root_secret**_ and **_root_**

```
mysql -u root -p'Cov04wxaA6zmkskNgrbDpv6DzKxTvgAJ' \
    -h testnc5-db  

```
and here I see e.g.:

```
MariaDB [(none)]> show databases;
+--------------------+
| Database           |
+--------------------+
| default            |
| information_schema |
| mysql              |
| performance_schema |
+--------------------+
4 rows in set (0.004 sec)

```

**Comment by james-cook** [comment]
It's not a user problem:

```
mysql -u nextcloud -p'1P2aMOLiulaii62puB3xX2YKRWTXXmfH' \
    -h testnc5-db  
```
(same command as above for the user "nextcloud" but without the "-D" database part ) does work.

It looks like an attempt is made to create the admin user on the database, but the database doesn't exist yet.

**Comment by james-cook** [comment]
Simply calling **./prepare_site.yml** a second time seems to clear up the problem.

UPDATE: well the error is somehow skipped, but the database is still missing:

As **root** db user:

```
MariaDB [(none)]> show databases;
+--------------------+
| Database           |
+--------------------+
| default            |
| information_schema |
| mysql              |
| performance_schema |
+--------------------+

```

Calling the "original" occ command in the **host** still shows the same error:

> docker exec --user www-data testnc5 php occ maintenance:install --database mysql --database-host testnc5-db --database-name nextcloud --database-table-prefix oc_ --database-user nextcloud --database-pass 1P2aMOLiulaii62puB3xX2YKRWTXXmfH --admin-user admin --admin-pass cVmLoyLJ09V9LJD7DIVtNr7UD81oYNl5 --data-dir /var/nc-data
> Error while trying to create admin user: Failed to connect to the database: An exception occurred in driver: SQLSTATE[HY000] [1044] Access denied for user 'nextcloud'@'%' to database 'nextcloud'

**Comment by james-cook** [comment]
DB log:
```
2020-02-06 12:09:37 22 [Warning] Access denied for user 'nextcloud'@'%' to database 'nextcloud'
2020-02-06 12:09:37 24 [Warning] Access denied for user 'nextcloud'@'%' to database 'nextcloud'
```

User nextcloud does exist:
```
MariaDB [(none)]> SELECT User, Host FROM mysql.user;
+-----------+-----------+
| User      | Host      |
+-----------+-----------+
| nextcloud | %         |
| root      | %         |
| root      | localhost |
+-----------+-----------+
3 rows in set (0.023 sec)

```

**Comment by james-cook** [comment]
The error only occurs with mysql.

By changing ONLY `app_db_type` in **_nextcloud.yml_** to postgres instead of mysql in a fresh droplet,  _**./prepare_site.yml**_ works.

It's really great to see the whole concept working (including traefik, portainer)! Thanks for a fantastic job :)

**And** it shows that the environment "in general" is OK. It really is just the mysql part that is failing.

Unfortunately for historical reasons I do need to get mysql working as my existing production nextclouds are all mysql.

**Comment by james-cook** [comment]
Running this again on a completely new droplet (just to check) - and I get the same error again with mysql.
The only workaround I can see at the moment is doing something like:

```
CREATE DATABASE nextcloud;
GRANT ALL PRIVILEGES on nextcloud.* to 'nextcloud' IDENTIFIED BY 'user-nextcloud-password';
```

in the playbook - before the **occ** call in the mysql case.

or manually before actually calling the playbook itself  (assumes you know the nextcloud user password). I did it after having called the playbook - which was a bit messy.
 

If you need to invoke the **occ** command yourself:

`docker exec --user www-data <nextcloud-container-name> php occ  maintenance:install  --database mysql --database-host <your-db-container> --database-name nextcloud  --database-table-prefix oc_ --database-user nextcloud  --database-pass <from-secrets-directory>  --admin-user admin --admin-pass <from-secrets-directory> --data-dir /var/nc-data `

**Comment by ReinerNippes** [comment]
I checked `docker logs <your-db-container>` and the database `default` is created instead of `nextcloud`. running the nextcloud_on_docker playbook on the same machine creates a `nextcloud` db. and it's nearly the same playbook task. 

@james-cook do you have your test machine running and could check with `docker inspect <your-db-container>` which ENV variables are set for the mysql container?

i have no clue what's going on. on the same machine nextcloud_on_docker run's fine. and it's nearly the playbook task to create the mysql db.

**Comment by james-cook** [comment]
Yes the database is `default` using this project and `nextcloud` on the nextcloud_on_docker.
This is the selfhosted_on_docker db env:

> docker exec -it testnc7-db sh
> $ env
> HOSTNAME=1e883504827c
> HOME=/home/mysql
> MYSQL_ROOT_PASSWORD=KvVFRhzLvj6mJg3AoXPDzDqwUmRv0a1E
> GPG_KEYS=177F4010FE56CA3336300305F1656F24C74CD1D8
> MARIADB_MAJOR=10.4
> TERM=xterm
> MYSQL_PASSWORD=B9QLS2GGPT7yBwDjvdr35CLfFUZeK4KT
> PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
> MYSQL_USER=nextcloud
> GOSU_VERSION=1.10
> MARIADB_VERSION=1:10.4.12+maria~bionic
> PWD=/
> MYSQL_DATABASE=default
> 

and this is the nextcloud_on_docker db env:

> docker exec -it nextcloud-db sh
> $ env
> HOSTNAME=701e4458e2a2
> HOME=/home/mysql
> MYSQL_ROOT_PASSWORD=Ay9Z7U2fed5DFSP3iDBYK5BX437pXlTI
> GPG_KEYS=177F4010FE56CA3336300305F1656F24C74CD1D8
> MARIADB_MAJOR=10.4
> TERM=xterm
> MYSQL_PASSWORD=XXXX
> PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
> MYSQL_USER=nextcloud
> GOSU_VERSION=1.10
> MARIADB_VERSION=1:10.4.12+maria~bionic
> PWD=/
> MYSQL_DATABASE=nextcloud
> 

I looked at config.yml and main.yml and `first setup nextcloud` in both looks very similar.

Ansible version on  nextcloud_on_docker: ansible 2.9.2
Ansible version on selfhosted_on_docker:ansible  2.9.4

I'm going to re-test with a completely "blank" Ubuntu 16.04 droplet.  
I may have not used a completely blank one in yesterday's tests and this may have had and influence.

**Comment by james-cook** [comment]
Using a new completely blank droplet I still get the same error :(  
The env inside the db is:

> $ env
> HOSTNAME=f6b882d02ccc
> HOME=/home/mysql
> MYSQL_ROOT_PASSWORD=N8tHTqjz6otQOQPpuEkcuisJFBKvTesA
> GPG_KEYS=177F4010FE56CA3336300305F1656F24C74CD1D8
> MARIADB_MAJOR=10.4
> TERM=xterm
> MYSQL_PASSWORD=1evQGcbcFjPXSbmJDwPDQfGKixz4XnoM
> PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
> MYSQL_USER=nextcloud
> GOSU_VERSION=1.10
> MARIADB_VERSION=1:10.4.12+maria~bionic
> PWD=/
> MYSQL_DATABASE=default
> 

Docker inspect on the db.ENV is the same:

>            "Env": [
>                 "MYSQL_PASSWORD=1evQGcbcFjPXSbmJDwPDQfGKixz4XnoM",
>                 "MYSQL_ROOT_PASSWORD=N8tHTqjz6otQOQPpuEkcuisJFBKvTesA",
>                 "MYSQL_USER=nextcloud",
>                 "MYSQL_DATABASE=default",
>                 "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
>                 "GOSU_VERSION=1.10",
>                 "GPG_KEYS=177F4010FE56CA3336300305F1656F24C74CD1D8",
>                 "MARIADB_MAJOR=10.4",
>                 "MARIADB_VERSION=1:10.4.12+maria~bionic"
>             ],
>         

At this point (as mentioned above) there is no `nextcloud` database in the database:

```
# docker exec -it nc1 sh
/var/www/html # apk add mysql-client
/var/www/html # mysql -u root -p'N8tHTqjz6otQOQPpuEkcuisJFBKvTesA' -h nc1-db
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 13
Server version: 10.4.12-MariaDB-1:10.4.12+maria~bionic mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> show databases;
+--------------------+
| Database           |
+--------------------+
| default            |
| information_schema |
| mysql              |
| performance_schema |
+--------------------+
4 rows in set (0.003 sec)

```

On my nextcloud_on_docker droplet I see:
```
MariaDB [(none)]> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| nextcloud          |
+--------------------+
2 rows in set (0.004 sec)

```

**Comment by james-cook** [comment]
This really does seem to be ansible regression - not fully tested but when I run the playbook with 2.9.2 the env appears to be correct and I don't get the error.

Caveat: this is an interim initial impression - the droplet is not necessarily "clean" as I was trying various ways to solve this problem. 

I'll spin up a new droplet with ansible 2.9.2 to check

**Comment by james-cook** [comment]
I tried downgrading ansible to 2.9.2 but this did not make a difference...

**Comment by ReinerNippes** [comment]
Found it. 

https://github.com/ReinerNippes/selfhosted_on_docker/blob/master/roles/docker_nextcloud/tasks/main.yml#L20

Postgresql & Nextcloud needs a database named `default`. The `nextcloud` db will created during the first setup run. Mariadb is different. The `nextcloud` database should be present. 

You may just change Line 20 and test it. Since you are only using mariadb that should be fine for you. 
I'll fix it over the weekend that it will work with both database engines.

**Comment by james-cook** [comment]
Thanks (as always :) ) for your quick response!

I can confirm that this does solve the problem.

One issue I did have was a "Failed to connect to database" on the second of two nextcloud "builds" (nc1 worked fine):
```
TASK [docker_nextcloud : nc2 | first setup nextcloud] ********************************************************************************************
Saturday 08 February 2020  15:44:24 +0000 (0:00:00.159)       0:04:23.441 *****
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "docker exec --user www-data nc2 php occ  maintenance:install  --database mysql --database-host nc2-db --database-name nextcloud  --database-table-prefix oc_ --database-user nextcloud  --database-pass BRcbsiBJuSgWHXngydEz7UvvVL2e9za7  --admin-user admin --admin-pass gPhfjxUdISJCLvsX0R61R74h23ModF1O --data-dir /var/nc-data\n", "delta": "0:00:02.049584", "end": "2020-02-08 15:44:26.823106", "msg": "non-zero return code", "rc": 1, "start": "2020-02-08 15:44:24.773522", "stderr": "", "stderr_lines": [], "stdout": "Error while trying to create admin user: Failed to connect to the database: An exception occurred in driver: SQLSTATE[HY000] [2002] Connection refused\n -> ", "stdout_lines": ["Error while trying to create admin user: Failed to connect to the database: An exception occurred in driver: SQLSTATE[HY000] [2002] Connection refused", " -> "]}

RUNNING HANDLER [prep_ufw : reload ufw] **********************************************************************************************************
Saturday 08 February 2020  15:44:26 +0000 (0:00:02.500)       0:04:25.941 *****

PLAY REC
```

Issuing the SAME `docker exec php occ...` command:
`docker exec --user www-data nc2 php occ  maintenance:install  --database mysql --database-host nc2-db --database-name nextcloud  --database-table-prefix oc_ --database-user nextcloud  --database-pass BRcbsiBJuSgWHXngydEz7UvvVL2e9za7  --admin-user admin --admin-pass gPhfjxUdISJCLvsX0R61R74h23ModF1O --data-dir /var/nc-data`

manually DID then work, so I wonder if we need to have a "wait on DB" task as you have in nextcloud_on_docker?

BTW: if you do ever have to issue the above **php occ...** command by hand, your config.php may not have been completely correctly assembled (contains only a few lines).

Simply re-running **./prepare_site.yml** will not fix this automatically.

You have to set `force_nextcloud_config: false` to `force_nextcloud_config: true` in _roles/docker_nextcloud/defaults/main.yml_ before calling **./prepare_site.yml** to ensure the config.php is rebuilt correctly.

**Comment by ReinerNippes** [comment]
hi james, 
you may test again. i hope i fixed both issues. 
regards.

**Comment by james-cook** [comment]
I tested this, and the results are very good. 

1. Mysql install fully works and 

2. the db "coming up" timeout ensures the build breaks off cleanly if there is a timeout.

How I tested:
- I updated from git, 
- merged in my specifics - 2 nextcloud sites, mysql db type etc.
- ensured DNS set-up correctly

Ran **./prepare_site.yml** and everything was fine (before I had had a failure on the second nc - see above - this error was later on in `first setup nextcloud` which left the install in an inconsistent state)

I then added 2 more nc instances in _selfhosted_sites.yml_, this time:

- one with OnlyOffice and 
- the other with collabora (my initial nextclouds do not have either).

The droplet is for testing only and **severely under-resourced** (1GB ram) but it does have 2GB swap...
In this way the setup is very good at testing this issue ;)

So, in fact on the second run now of **./prepare_site.yml** I did get a timeout:

```
TASK [docker_nextcloud : wait for db container to become ready] **********************************************************************************
Tuesday 11 February 2020  12:15:23 +0000 (0:00:00.856)       0:07:00.031 ******
fatal: [localhost]: FAILED! => {"changed": false, "elapsed": 30, "msg": "Timeout when waiting for 172.22.0.3:3306"}

PLAY RECAP ***************************************************************************************************************************************
localhost                  : ok=147  changed=38   unreachable=0    failed=1    skipped=49   rescued=0    ignored=0
```

Because of now **breaking off cleanly** rather than continuing - this problem could be easily solved just by **re-running the playbook**.

A much better situation.

**Comment by Nickous** [comment]
I can confirm you that you solved the problem about creating nextcloud mysql database.
I just had to increase the timeout to 60 because my VPS server must bo too slow :)

**Comment by Nickous** [comment]
@james-cook Do you succeed in installing and using OnlyOffice.
With version Nextcloud18, there is now a native OnlyOffice app. 
But when I try to install "Community Document Server" app, I've got this error message
> cURL error 28: Operation timed out after 30000 milliseconds with 222731572 out of 315849196 bytes received (see http://curl.haxx.se/libcurl/c/libcurl-errors.html) 

If I use the @ReinerNippes OnlyOffice install (`onlyoffice:  true` from file `selfhosted_sites.yml`), the onlyoffice docker is well built but no Opendocument available in my Nextcloud.
What I have to enter in the `Document Editing Service address` parameter in ONLYOFFICE settings administration?

**Comment by james-cook** [comment]
@Nickous I would open a separate issue for onlyoffice/newer versions...

**Comment by ReinerNippes** [comment]
@Nickous I couldn't setup the new onlyoffice app integrated in NC18 yet. I tried with nextcloud_on_docker and got up to now only errors. I guess it's due to the reverse proxy settings. But had no time to dive into it. 

`onlyoffice: true` do you use selfsigned certs? If not I have to check the playbook.

**Comment by james-cook** [comment]
@ReinerNippes  IMHO the specific issue - nextcloud mysql not installed/handled cleanly is solved - should we close this issue?

**Comment by ReinerNippes** [comment]
@james-cook yes. also [the problem when you create a second nc instance](https://github.com/ReinerNippes/selfhosted_on_docker/issues/3#issuecomment-584617077)? or did i get you wrong?

**Comment by james-cook** [comment]
@ReinerNippes yes I'll close both (other issue was for nextcloud_on_docker)
